### ISSUE TITLE
Since pydantic 1.10.3 dataclasses with `__setattr__` defined cause an exception

### ISSUE THREAD

**Issue by samuelcolvin** [pydantic/pydantic#4907]
### Discussed in https://github.com/pydantic/pydantic/discussions/4905

<div type='discussions-op-text'>

<sup>Originally posted by **KramNamez** January  4, 2023</sup>
As seen in https://github.com/nabla-c0d3/sslyze/issues/586 SSLyze broke, despite pydantic only doing point releases that should not change its API.

Freezing pydantic at 1.10.2 fixes the issue for now, but I still wanted to ask if this is a bug in pydantic or in SSLyze's usage of pydantic.</div>

In short, there's another edge case to dataclasses related to `__setattr__` and probably #4880.

**Comment by samuelcolvin** [comment]
Questions:
* is this specific to python 3.7?
* could we work around it by checking for `__setattr__` before defining it?

https://github.com/pydantic/pydantic/blob/63c683d9fc7b494bf81e964f97ec2f9b78b8e09d/pydantic/dataclasses.py#L257-L258

**Comment by KramNamez** [comment]
No, I'm using Python 3.10 - this is not specific to Python 3.7. (Unfortunately, as that would've been a relatively easy fix for me, at least.)

The code which seems to trigger this is here: https://github.com/nabla-c0d3/sslyze/blob/412d08f62695622b152374777591278702f5d651/sslyze/plugins/compression_plugin.py#L24

At a quick glance, I don't see where this uses `__setattr__`, but I'll dig some more.

**Comment by samuelcolvin** [comment]
It might be that the parent type uses `ABC`.

**Comment by KramNamez** [comment]
Indeed it does!

**Comment by PrettyWood** [comment]
I reckon it's more an issue with frozen dataclasses than `ABC`. I'll check tonight

**Comment by PrettyWood** [comment]
Ok I checked now. Yes that's very weird. If `CompressionScanResult` has no parent class (abc or not) this works.
But I guess the easiest solution is to do
```py
CompressionScanResultAsJson = pydantic.dataclasses.dataclass(CompressionScanResult)  # no need to add frozen here
```
It will still be frozen and avoid the error.
I will dig deeper tonight but as you can see it seems to be a (yet again) weird edge case

**Comment by samuelcolvin** [comment]
Thanks so much for looking into this @PrettyWood and sorry to waste your time.

**Comment by KramNamez** [comment]
Thank you so much for the analysis. I'll see that I can get a fix submitted to SSLyze to resolve this!
